11/debian-9/rootfs/libpostgresql.sh

```diff
@@ -0,0 +1,694 @@
+    postgresql_set_property "wal_keep_segments" "12"
+    postgresql_set_property "hot_standby" "on"
+    if (( POSTGRESQL_NUM_SYNCHRONOUS_REPLICAS > 0 ));then
+        postgresql_set_property "synchronous_commit" "$POSTGRESQL_SYNCHRONOUS_COMMIT_MODE"
+        postgresql_set_property "synchronous_standby_names" "$POSTGRESQL_NUM_SYNCHRONOUS_REPLICAS ($POSTGRESQL_CLUSTER_APP_NAME)"
+    fi
+}
+
+########################
+# Alter password of the postgres user
+# Globals:
+#   POSTGRESQL_*
+# Arguments:
+#   None
+# Returns:
+#   None
+#########################
+postgresql_alter_postgres_user() {
+    local -r escaped_password="${POSTGRESQL_PASSWORD//\'/\'\'}" 
+    info "Changing password of ${POSTGRESQL_USERNAME}"
+    echo "ALTER ROLE postgres WITH PASSWORD '$escaped_password';" | postgresql_execute
+}
+
+########################
+# Create an admin user with all privileges in POSTGRESQL_DATABASE
+# Globals:
+#   POSTGRESQL_*
+# Arguments:
+#   None
+# Returns:
+#   None
+#########################
+postgresql_create_admin_user() {
+    local -r escaped_password="${POSTGRESQL_PASSWORD//\'/\'\'}" 
+    info "Creating user ${POSTGRESQL_USERNAME}"
+    echo "CREATE ROLE ${POSTGRESQL_USERNAME} WITH LOGIN CREATEDB PASSWORD '${escaped_password}';" | postgresql_execute
+    info "Grating access to ${POSTGRESQL_USERNAME} to the database ${POSTGRESQL_DATABASE}"
+    echo GRANT ALL PRIVILEGES ON DATABASE "${POSTGRESQL_DATABASE}" TO "${POSTGRESQL_USERNAME}"\; | postgresql_execute "" "postgres" "$POSTGRESQL_PASSWORD"
+}
+
+########################
+# Create a database with name $POSTGRESQL_DATABASE
+# Globals:
+#   POSTGRESQL_*
+# Arguments:
+#   None
+# Returns:
+#   None
+#########################
+postgresql_create_custom_database() {
+    echo "CREATE DATABASE $POSTGRESQL_DATABASE" | postgresql_execute "" "postgres" "" "localhost"
+}
+
+########################
+# Change postgresql.conf to listen in 0.0.0.0
+# Arguments:
+#   None
+# Returns:
+#   None
+#########################
+postgresql_enable_remote_connections() {
+    postgresql_set_property "listen_addresses" "*"
+}
+
+########################
+# Check if a given configuration file was mounted externally
+# Globals:
+#   POSTGRESQL_*
+# Arguments:
+#   $1 - Filename
+# Returns:
+#   1 if the file was mounted externally, 0 otherwise
+#########################
+postgresql_is_file_external() {
+    local -r filename=$1
+    if [[ -d "$POSTGRESQL_MOUNTED_CONF_DIR" ]] && [[ -f "$POSTGRESQL_MOUNTED_CONF_DIR"/"$filename" ]]; then
+        return 0
+    else
+        return 1
+    fi
+}
+
+########################
+# Ensure PostgreSQL is initialized
+# Globals:
+#   POSTGRESQL_*
+# Arguments:
+#   None
+# Returns:
+#   None
+#########################
+postgresql_initialize() {
+    info "Initializing PostgreSQL database..."
+
+    # User injected custom configuration
+    if [[ -d "$POSTGRESQL_MOUNTED_CONF_DIR" ]] && compgen -G "$POSTGRESQL_MOUNTED_CONF_DIR"/* > /dev/null;then
```

Comment:
what is `compgen` and how does it work? I understand you are just checking that $POSTGRESQL_MOUNTED_CONF_DIR exists and is not empty?

Reply:
compgen is for checking that there are files inside the folder, otherwise the cp /* command fails

Reply:
ok